### PR TITLE
Specs for File.open('x'), exception: false, KeyError constructor options.

### PR DESCRIPTION
**Pull request by iliabylich**
+ File.open with 'x' flag.
+ `Kernel#Complex` with exception: false.
+ `Kernel#Float` with exception: false.
+ `Kernel#Integer` with exception: false.
+ `Kernel#Rational` with exception: false.
+ new `:receiver` and `:key` options for `KeyError`.

### CODE REVIEW

**Comment by iliabylich** [comment]
@eregon Could you take a look please?

File: core/kernel/Complex_spec.rb [hunk 1]
```diff
@@ -138,4 +138,50 @@
...
+
+      describe "and [non-Numeric]" do
+        it "swallows an error" do
+          Complex(:sym, exception: false).should == nil
+        end
+      end
+
+      describe "and [non-Numeric, Numeric] argument" do
+        it "throws a TypeError" do
+          lambda { Complex(:sym, 0, exception: false) }.should raise_error(TypeError, "not a real")
```
Comment (eregon) [inline core/kernel/Complex_spec.rb]: This sounds like a bug, could you report it to MRI?

File: core/kernel/Complex_spec.rb [hunk 2]
```diff
@@ -138,4 +138,50 @@
...
+          Complex("a", 0, exception: false).should == nil
+          Complex(0, "b", exception: false).should == nil
+        end
+      end
+
+      describe "and nil arguments" do
+        it "still throws a TypeError" do
+          lambda { Complex(nil, exception: false) }.should raise_error(TypeError, "can't convert nil into Complex")
+          lambda { Complex(0, nil, exception: false) }.should raise_error(TypeError, "can't convert nil into Complex")
+          lambda { Complex(nil, 0, exception: false) }.should raise_error(TypeError, "can't convert nil into Complex")
```
Comment (eregon) [inline core/kernel/Complex_spec.rb]: This also sounds buggy, could you report it to MRI?

Comment (eregon) [reply core/kernel/Complex_spec.rb]: Especially since `Float(nil, exception: false)` returns nil.

File: core/kernel/Integer_spec.rb [hunk 3]
```diff
@@ -99,6 +99,51 @@
...
+        it "raises a TypeError" do
+          lambda { Integer(infinity_value, exception: false) }.should raise_error(FloatDomainError)
+        end
+      end
+
+      describe "and passed nil" do
+        it "swallows an error" do
+          Integer(nil, exception: false).should == nil
+        end
+      end
```
Comment (eregon) [inline core/kernel/Integer_spec.rb]: Could add a spec where it's a valid input, like `"123"` and one where it's an invalid String, like `"abc"`?

File: shared/rational/Rational.rb [hunk 4]
```diff
@@ -100,4 +100,44 @@
...
+          Rational("a", "b", exception: false).should == nil
+          Rational("a", 0, exception: false).should == nil
+          Rational(0, "b", exception: false).should == nil
+        end
+      end
+
+      describe "and nil arguments" do
+        it "raises a TypeError" do
+          lambda { Rational(nil, exception: false) }.should raise_error(TypeError, "can't convert nil into Rational")
+          lambda { Rational(nil, nil, exception: false) }.should raise_error(TypeError, "can't convert nil into Rational")
```
Comment (eregon) [inline shared/rational/Rational.rb]: This also sounds buggy, could you report it to MRI?

**Comment by eregon** [comment]
@iliabylich Looks great. Could you add the spec as mentioned in the comments, and report the potential bugs/weird behavior to the MRI issue tracker?

**Comment by iliabylich** [comment]
> Could add a spec where it's a valid input, like "123" and one where it's an invalid String, like "abc"?

Done

> report the potential bugs/weird behavior to the MRI issue tracker

Also done - https://bugs.ruby-lang.org/issues/15525

**Pull request merged**